controller/kongplugininstallation/controller.go

```diff
@@ -0,0 +1,216 @@
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.GetLogger(ctx, "kongplugininstallation", r.DevelopmentMode)
+
+	log.Trace(logger, "reconciling KongPluginInstallation resource", req)
+	var kpi v1alpha1.KongPluginInstallation
+	if err := r.Client.Get(ctx, req.NamespacedName, &kpi); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Trace(logger, "managing KongPluginInstallation resource", kpi)
+	var credentialsStore credentials.Store
+	if kpi.Spec.ImagePullSecretRef != nil {
+		secretNN := client.ObjectKey{
+			Namespace: kpi.Spec.ImagePullSecretRef.Namespace,
+			Name:      kpi.Spec.ImagePullSecretRef.Name,
+		}
+		if secretNN.Namespace == "" {
+			secretNN.Namespace = req.Namespace
+		}
+
+		var secret corev1.Secret
+		if err := r.Client.Get(
+			ctx,
+			secretNN,
+			&secret,
+		); err != nil {
+			return ctrl.Result{}, setStatusConditionFailed(ctx, r.Client, &kpi, fmt.Sprintf("cannot retrieve secret %q, because: %s", secretNN, err))
+		}
+
+		const requiredKey = ".dockerconfigjson"
+		secretData, ok := secret.Data[requiredKey]
+		if !ok {
+			return ctrl.Result{}, setStatusConditionFailed(
+				ctx, r.Client, &kpi, fmt.Sprintf("can't parse secret %q - unexpected type, it should follow 'kubernetes.io/dockerconfigjson'", secretNN),
+			)
+		}
+		var err error
+		credentialsStore, err = image.CredentialsStoreFromString(string(secretData))
+		if err != nil {
+			return ctrl.Result{}, setStatusConditionFailed(ctx, r.Client, &kpi, fmt.Sprintf("can't parse secret: %q data: %s", secretNN, err))
+		}
+	}
+
+	plugin, err := image.FetchPluginContent(ctx, kpi.Spec.Image, credentialsStore)
+	if err != nil {
+		return ctrl.Result{}, setStatusConditionFailed(ctx, r.Client, &kpi, fmt.Sprintf("problem with the image: %q error: %s", kpi.Spec.Image, err))
+	}
+
+	cms, err := kubernetes.ListConfigMapsForOwner(ctx, r.Client, kpi.GetUID())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	var cm corev1.ConfigMap
+	switch len(cms) {
+	case 0:
+		if cmName := kpi.Status.UnderlyingConfigMapName; cmName != "" {
+			cm.Name = cmName
+		} else {
+			cm.GenerateName = kpi.Name
+		}
+		cm.Namespace = kpi.Namespace
+		cm.Data = map[string]string{
+			fmt.Sprintf("%s.lua", kpi.Name): string(plugin),
+		}
+		if err := ctrl.SetControllerReference(&kpi, &cm, r.Scheme); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.Client.Create(ctx, &cm); err != nil {
+			return ctrl.Result{}, err
+		}
+		kpi.Status.UnderlyingConfigMapName = cm.Name
+	case 1:
+		cm = cms[0]
+		cm.Name = kpi.Status.UnderlyingConfigMapName
+		cm.Data = map[string]string{
+			fmt.Sprintf("%s.lua", kpi.Name): string(plugin),
+		}
+		if err := r.Client.Update(ctx, &cm); err != nil {
+			return ctrl.Result{}, err
+		}
+	default:
+		// It should never happen.
+		return ctrl.Result{}, errors.New("unexpected error happened - more than one ConfigMap found")
```

Comment:
What about this one? What do you think?

Reply:
Good question. I think that it can only happen when someone manually creates such ConfigMap and sets its owner to KPI, that more than one ConfigMap will be here available... On the other hand how to decide which one I want in such case 🤔

Reply:
ok, let's defer this one and implement in case we need it :)